amaranth/vendor/gowin.py

```diff
@@ -0,0 +1,549 @@
+            {{emit_rtlil()}}
+        """,
+        "{{name}}.ys": r"""
+            # {{autogenerated}}
+            {% for file in platform.iter_files(".v") -%}
+                read_verilog {{get_override("read_verilog_opts")|options}} {{file}}
+            {% endfor %}
+            {% for file in platform.iter_files(".sv") -%}
+                read_verilog -sv {{get_override("read_verilog_opts")|options}} {{file}}
+            {% endfor %}
+            {% for file in platform.iter_files(".il") -%}
+                read_ilang {{file}}
+            {% endfor %}
+            read_ilang {{name}}.il
+            delete w:$verilog_initial_trigger
+            {{get_override("script_after_read")|default("# (script_after_read placeholder)")}}
+            synth_gowin {{get_override("synth_opts")|options}} -top {{name}} -json {{name}}.json
+            {{get_override("script_after_synth")|default("# (script_after_synth placeholder)")}}
+        """,
+    }
+
+    _apicula_command_templates = [
+        r"""
+        {{invoke_tool("yosys")}}
+            {{quiet("-q")}}
+            {{get_override("yosys_opts")|options}}
+            -l {{name}}.rpt
+            {{name}}.ys
+        """,
+        r"""
+        {{invoke_tool("nextpnr-gowin")}}
+            {{quiet("--quiet")}}
+            {{get_override("nextpnr_opts")|options}}
+            --log {{name}}.tim
+            --device {{platform._part}}
+            --family {{platform.device}}
+            --json {{name}}.json
+            --freq 27
+            --cst {{name}}.cst
+            --write {{name}}.json
+        """,
+        r"""
+        {{invoke_tool("gowin_pack")}}
+            {{verbose("-v")}}
+            -d {{platform.device}}
+            -o {{name}}.fs
+            {{get_override("gowin_pack_opts")|options}}
+            {{name}}.json
+        """
+    ]
+
+    ### Gowin vendor toolchain templates
+
+    _gowin_required_tools = ["gw_sh"]
+
+    @property
+    def _device_name(self):
+        return "{}-{}{}".format(self.series, self.size, self.subseries)
+
+    _gowin_file_templates = {
+        **TemplatedPlatform.build_script_templates,
+        **_common_file_templates,
+        "{{name}}.tcl": r"""
+            # {{autogenerated}}
+            {% for file in platform.iter_files(".v",".sv",".vhd",".vhdl") -%}
+                add_file {{file}}
+            {% endfor %}
+            add_file -type verilog {{name}}.v
+            add_file -type cst {{name}}.cst
+            add_file -type sdc {{name}}.sdc
+            set_device -name {{platform._device_name}} {{platform._part}}
+            set_option -verilog_std v2001 -print_all_synthesis_warning 1 -show_all_warn 1
+            {{get_override("add_options")|default("# (add_options placeholder)")}}
+            run all
+        """,
+    }
+
+    _gowin_command_templates = [
+        r"""
+        {{invoke_tool("gw_sh")}}
+            {{name}}.tcl
+        """,
+        "cp impl/pnr/project.fs {{name}}.fs"
```

Comment:
This will not work on Windows. Does Gowin ship Windows versions of the tools?

Reply:
Yes, gowin EDA also exists for Windows. Does a template exist already for such a copy operation?
I'll try to put that copy operation into the tcl file that is generated. That ought to be platform independent.

Reply:
No, a template does not exist. I was thinking it could be introduced, but your Tcl solution seems good!